src/bundler.ts

```diff
@@ -102,346 +97,268 @@ export class Bundler {
-    if (this.isBlankTextNode(next)) {
-      dom5.remove(next);
-    }
-    if (replacement) {
-      dom5.replace(node, replacement);
+  /**
+   * Set the hidden div at the appropriate location within the document.  The
+   * goal is to place the hidden div at the same place as the first html
+   * import.  However, the div can't be placed in the `<head>` of the document
+   * so if first import is found in the head, we prepend the div to the body.
+   * If there is no body, we'll just attach the hidden div to the document at
+   * the end.
+   */
+  private _attachHiddenDiv(document: ASTNode, hiddenDiv: ASTNode) {
+    const firstHtmlImport = dom5.query(document, matchers.htmlImport);
+    const body = dom5.query(document, matchers.body);
+    if (body) {
+      if (firstHtmlImport &&
+          dom5.predicates.parentMatches(matchers.body)(firstHtmlImport)) {
+        astUtils.insertAfter(firstHtmlImport, hiddenDiv);
+      } else {
+        astUtils.prepend(body, hiddenDiv);
+      }
     } else {
-      dom5.remove(node);
+      dom5.append(document, hiddenDiv);
     }
   }
 
-  isLicenseComment(node: ASTNode): boolean {
-    if (dom5.isCommentNode(node)) {
-      return dom5.getTextContent(node).indexOf('@license') > -1;
+  /**
+   * Produces a document containing the content of all of the bundle's files.
+   * If the bundle's url resolves to an existing html file, that file will be
+   * used as the basis for the generated document.
+   */
+  private async _bundleDocument(
+      docBundle: AssignedBundle,
+      bundleManifest: BundleManifest,
+      bundleImports?: Set<string>): Promise<ASTNode> {
+    const docUrl = docBundle.url;
+    const document = await this._prepareBundleDocument(docBundle);
+    this._appendHtmlImportsForBundle(document, docBundle);
+    importUtils.rewriteImportedUrls(this.basePath, document, docUrl, docUrl);
+
+    await this._inlineHtmlImports(docUrl, document, docBundle, bundleManifest);
+
+    if (this.enableScriptInlining) {
+      await this._inlineScripts(docUrl, document);
+    }
+    if (this.enableCssInlining) {
+      await this._inlineStylesheetLinks(docUrl, document);
+      await this._inlineStylesheetImports(docUrl, document);
+    }
+
+    if (this.stripComments) {
+      astUtils.stripComments(document);
     }
-    return false;
+    return document;
   }
 
   /**
    * Creates a hidden container <div> to which inlined content will be
    * appended.
    */
-  createHiddenContainerNode(): ASTNode {
+  private _createHiddenDiv(): ASTNode {
     const hidden = dom5.constructors.element('div');
     dom5.setAttribute(hidden, 'hidden', '');
     dom5.setAttribute(hidden, 'by-polymer-bundler', '');
     return hidden;
   }
 
   /**
-   * Inline external scripts <script src="*">
+   * Given an array of Bundles, remove all files from bundles which are in the
+   * "excludes" set.  Remove any bundles which are left empty after excluded
+   * files are removed.
```

Comment:
Maybe note that while you mutate the Bundles, you don't mutate the bundles array. Would eventually be nice if this were a pure function, maybe if Bundle had a copy constructor.